caom2pipe/validator_composable.py

```diff
@@ -0,0 +1,269 @@
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+    def _filter_result(self, dest_meta_temp):
+        """The default implementation does nothing, but this allows
+        implementations that extend this class to remove entries from
+        the comparison results for whatever reason might come up."""
+        pass
+
+    def _find_unaligned_dates(self, source, data):
+        newer = pd.DataFrame()
+        if len(data) > 0:
+            # SG - 08-09-22 - All timestamps in the SI databases are in UT.
+            #
+            # source_temp => url, timestamp, f_name
+            # data => uri, contentLastModified
+            data['f_name'] = data.uri.apply(Validator.filter_meta)
+            data['ts'] = data.contentLastModified.apply(pd.to_datetime)
+            merged = pd.merge(source, data, how='inner', on=['f_name'])
+            newer = merged[merged.timestamp > merged.ts]
+        return newer
+
+    def _read_list_from_destination_data(self):
+        """Code to execute a query for files and the arrival time, that are in
+        CADC storage.
+        """
+        query = (
+            f"SELECT A.uri, A.contentLastModified "
+            f"FROM inventory.Artifact AS A "
+            f"WHERE A.uri LIKE '%:{self._config.collection}/%' "
+            f"AND A.uri NOT LIKE '%{self._preview_suffix}'"
+        )
+        self._logger.debug(f'Query is {query}')
+        temp = query_tap_pandas(query, self._data_client)
+        temp['f_name'] = temp.uri.apply(Validator.filter_meta)
+        # columns are uri, contentLastModified, f_name
+        return temp
+
+    def _read_list_from_destination_meta(self):
+        query = (
+            f"SELECT O.observationID, A.uri FROM caom2.Observation AS O "
+            f"JOIN caom2.Plane AS P ON O.obsID = P.obsID "
+            f"JOIN caom2.Artifact AS A ON P.planeID = A.planeID "
+            f"WHERE O.collection='{self._config.collection}' "
+            f"AND A.uri not like '%{self._preview_suffix}'"
+        )
+        self._logger.debug(f'Query is {query}')
+        temp = query_tap_pandas(query, self._caom_client)
+        temp['f_name'] = temp.uri.apply(Validator.filter_meta)
+        # columns are observationID, uri, f_name
+        return temp
+
+    def read_from_source(self):
+        """Read the entire source site listing. This function is expected to
+        return a dict of all the file names available from the source, where
+        the keys are the file names, and the values are the timestamps at the
+        source."""
+        raise NotImplementedError()
+
+    def validate(self):
+        # META
+        self._logger.info('Query destination metadata.')
+        dest_meta_temp = self._read_list_from_destination_meta()
+
+        # SOURCE
+        self._logger.info('Query source metadata.')
+        source_temp = self.read_from_source()
+
+        # SOURCE vs META
+        self._logger.info('Find files that do not appear at CADC.')
```

Comment:
If I understand correctly, we synchronize remote data with CADC storage. This synchronization assumes metadata extraction. So at the end of this step, all 3 repos (SOURCE, DATA and META) should have the same info. Why do we need the META-SOURCE validation? We might need META-DATA validation just to prevent some potential errors but in general, if the pipeline works correctly they should be in sync and I would think that the step is just our internal maintenance. It's the script that Séverin or Alinga used to run periodically. So I don't know if this step is necessary in the pipeline (unless I'm misunderstanding the design)

Reply:
You are not mis-understanding the design. The META portion of this comparison is going to be removed.